# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [10]:
# Importar pandas
import pandas as pd


Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [11]:
# Leer el archivo y almacenarlo en df
df = pd.read_csv ("C:\\Users\\HP\\Downloads\\music_project_en.csv")


Muestra las 10 primeras filas de la tabla:

In [12]:
# Obtener las 10 primeras filas de la tabla df
df = pd.read_csv("C:\\Users\\HP\\Downloads\\music_project_en.csv")
print(df.head())


     userID                        Track            artist  genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile   rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg   rock   
2    20EC38            Funiculì funiculà       Mario Lanza    pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice   folk   
4  E2DC1FAE                  Soul People        Space Echo  dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  


Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [13]:
# Obtener la información general sobre nuestros datos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. `Detecta el tercer problema por tu cuenta y descríbelo aquí`.
"UserID" no se encuentra con la estructura snake_case donde se se separan las palabras con guiones bajos




### Escribe observaciones de tu parte. Estas son algunas de las preguntas que pueden ser útiles: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`Segun la funcion la funcion info nos dice que tenemos 65079 filas del 0 al 65078, algunas columnas tienen valores no nulos que estan completos que no son nulo como en algunas columnas y las columnas son de tipo objeto son de cadena de texto  

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?` se necesita hacer el preprocesamiento de los datos, analizar los datos ausentes de las columnas

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`Solo con la funcion info se muestran que algunas columnas hay valores ausentes, y no hay datos incorrectos pero mas adelantes se pueden presentar

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [14]:
# Muestra los nombres de las columnas
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

Anteriormente, aprendiste acerca de la forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [15]:
# Bucle en los encabezados poniendo todo en minúsculas
new_columns = []
for column in df.columns:
    name_lowered = column.lower()

    new_columns.append(name_lowered)
df.columns = new_columns
print(df.columns)


Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')


Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [16]:
# Bucle en los encabezados eliminando los espacios
new_columns = []
for column in df.columns:
    old_name = column.strip()
    new_columns.append(old_name)
df.columns = new_columns
print(df.columns)

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [17]:
# Cambiar el nombre de la columna "userid"
new_columns= []
for old_name in df.columns:
    name_stripped = old_name.strip()
    name_lowered = name_stripped.lower()
    name_no_spaces = name_lowered.replace(' ', '_')
    new_columns.append(name_no_spaces)
df.columns = new_columns
df.rename(columns={'userid': 'user_id'}, inplace=True)
print(df.columns)


Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Comprueba el resultado. Muestra los encabezados una vez más:

In [18]:
# Comprobar el resultado: la lista de encabezados
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [19]:
# Calcular el número de valores ausentes
print(df.isna().sum())


user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [38]:
# Lista de columnas a las que se les reemplazarán los valores ausentes
columns_to_replace = ['track', 'artist', 'genre']

# Bucle para reemplazar los valores ausentes con 'unknown'
for column in columns_to_replace:
    df[column] = df[column].fillna('unknown')

# Mostrar los valores ausentes después del reemplazo
print('Valores ausentes después del reemplazo:')
print(df[columns_to_replace].isna().sum())

    

Valores ausentes después del reemplazo:
track     0
artist    0
genre     0
dtype: int64


Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [21]:
# Contar valores ausentes
missing_values_replacement = df.isnull().sum()
print('Numero de valores ausentes despues del reemplazo')
print(missing_values_replacement)

Numero de valores ausentes despues del reemplazo
user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [22]:
# Contar duplicados explícitos
num_duplicates = df.duplicated().sum()
print(num_duplicates)

3826


Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [23]:
# Eliminar duplicados explícitos
df.drop_duplicates(inplace=True)
print(df.head())

    user_id                        track            artist  genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile   rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg   rock   
2    20EC38            Funiculì funiculà       Mario Lanza    pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice   folk   
4  E2DC1FAE                  Soul People        Space Echo  dance   

          city      time        day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  


Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [24]:
# Comprobar de nuevo si hay duplicados
df.drop_duplicates(inplace=True)
num_duplicates_after = df.duplicated().sum()
print('Numero de duplicados despues de eliminar:', num_duplicates_after)


Numero de duplicados despues de eliminar: 0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [25]:
# Inspeccionar los nombres de géneros únicos
unique_genres = df['genre'].unique()
sorted_genres = sorted(unique_genres)
print(sorted_genres)
print()

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [26]:
# Función para reemplazar duplicados implícitos
def replace_wrong_genres(df, correct_genre, wrong_genres):
    for wrong_genre in wrong_genres: 
     df['genre'] = df['genre'].replace(wrong_genre, correct_genre)
    return df

Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [27]:
# Eliminar duplicados implícitos
def replace_wrong_genres(df, correct_genre, wrong_genres):
    for wrong_genre in wrong_genres: 
     df['genre'] = df['genre'].str.replace(wrong_genre, correct_genre, case= False, regex=False)
     return df
wrong_genres = ['hip', 'hop', 'hip-hop']
correct_genre = 'hiphop'
df = replace_wrong_genres(df, correct_genre, wrong_genres)
print(df['genre'].unique())

['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hiphop' 'jazz' 'postrock'
 'latin' 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental'
 'rusrock' 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'hiphophop' 'drum' 'extrememetal' 'türkçe'
 'experimental' 'easy' 'metalcore' 'modern' 'argentinetango' 'old' 'swing'
 'breaks' 'euro

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [28]:
# Comprobación de duplicados implícitos
def replace_wrong_genres(df, correct_genre, wrong_genres):
    for wrong_genre in wrong_genres: 
     df['genre'] = df['genre'].str.replace(wrong_genre, correct_genre, case= False, regex=False)
     return df
wrong_genres = ['hip', 'hop', 'hip-hop']
correct_genre = 'hiphop'
df = replace_wrong_genres(df, correct_genre, wrong_genres)
unique_genres = df['genre'].unique()
print(unique_genres)

['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hiphophop' 'jazz' 'postrock'
 'latin' 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental'
 'rusrock' 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'hiphophophop' 'drum' 'extrememetal' 'türkçe'
 'experimental' 'easy' 'metalcore' 'modern' 'argentinetango' 'old' 'swing'
 'breaks'

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

`Describe brevemente lo que has notado al analizar duplicados, cómo abordaste sus eliminaciones y qué resultados obtuviste.` En esta parte de preprocesamiento de los datos de identificaron los valores ausentes y se procedio a reemplazarlos, para los datos duplicados se contaron, se eliminaron, pero donde podria influir era en la columna 'genre' se formulo una funcion para poder eliminar generos duplicados y finalmente devolver los valores unicos de la columna 'genre'

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [29]:
# Contar las canciones reproducidas en cada ciudad

specific_days = ['Monday', 'Wednesday', 'Friday']
filtered_df = df[df['day'].isin(specific_days)]


count_monday = filtered_df[filtered_df['day'] == 'Monday'].groupby('city')['track'].count()
count_wednesday = filtered_df[filtered_df['day'] == 'Wednesday'].groupby('city')['track'].count()
count_friday = filtered_df[filtered_df['day'] == 'Friday'].groupby('city')['track'].count()

print('Canciones reproducidas el lunes:')                           
print(count_monday)
print('Canciones reproducidas el miercoles:') 
print(count_wednesday)
print('Canciones reproducidas el viernes:') 
print(count_friday)

Canciones reproducidas el lunes:
city
Shelbyville     5614
Springfield    15740
Name: track, dtype: int64
Canciones reproducidas el miercoles:
city
Shelbyville     7003
Springfield    11056
Name: track, dtype: int64
Canciones reproducidas el viernes:
city
Shelbyville     5895
Springfield    15945
Name: track, dtype: int64


`Comenta tus observaciones aquí` Aqui si afirmariamos nuestra hipotesis se muestran las canciones reproducidas los dias lunes, miercoles y viernes de ambas ciudad y nos muestran que Springfield reproduce mas canciones aunque no es un diferencia tan marcada

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [30]:
# Calcular las canciones reproducidas en cada uno de los tres días
specific_days = ['Monday', 'Wednesday', 'Friday']
filtered_df = df[df['day'].isin(specific_days)]
count_by_day = filtered_df.groupby('day')['track'].count()
print('Numero de canciones reproducidas en cada uno de los tres dias:')
print(count_by_day)

Numero de canciones reproducidas en cada uno de los tres dias:
day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64


`Comenta tus observaciones aquí` estos datos no nos ayudarian a rechzar o aceptar nuestra hipotesis ya que se mezclan ambas ciudades sobre el numero de canciones reproducidas por cada uno de los tres dias, nos ayudarian datos que nos muestren por ciudad y por dia

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [31]:
# Declara la función number_tracks() con dos parámetros: day= y city=.
def number_tracks(day,city):

    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    filtered_by_day = df[df['day'] == day]

    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    filtered_by_city = filtered_by_day[filtered_by_day['city'] == city]

    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    track_count = filtered_by_city['user_id'].count()
    

    # Devolve el número de valores de la columna 'user_id'
    return track_count
monday_springfield_tracks = number_tracks('Monday', 'Springfield')
print(f'Numero de canciones reproducidas en Springfield el lunes : {monday_springfield_tracks}')
monday_shelbyville_tracks = number_tracks('Monday', 'Shelbyville')
print(f'Numero de canciones reproducidas en Shelbyville el lunes : {monday_shelbyville_tracks}')
wednesday_springfield_tracks = number_tracks('Wednesday', 'Springfield')
print(f'Numero de canciones reproducidas en Springfield el miercoles : {wednesday_springfield_tracks}')
wednesday_shelbyville_tracks = number_tracks('Wednesday', 'Shelbyville')
print(f'Numero de canciones reproducidas en Shelbyville el miercoles : {wednesday_shelbyville_tracks}')
friday_springfield_tracks = number_tracks('Friday', 'Springfield')
print(f'Numero de canciones reproducidas en Springfield el viernes : {friday_springfield_tracks}')
friday_shelbyville_tracks = number_tracks('Friday', 'Shelbyville')
print(f'Numero de canciones reproducidas en Shelbyville el viernes : {friday_shelbyville_tracks}')



Numero de canciones reproducidas en Springfield el lunes : 15740
Numero de canciones reproducidas en Shelbyville el lunes : 5614
Numero de canciones reproducidas en Springfield el miercoles : 11056
Numero de canciones reproducidas en Shelbyville el miercoles : 7003
Numero de canciones reproducidas en Springfield el viernes : 15945
Numero de canciones reproducidas en Shelbyville el viernes : 5895


Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [32]:
# El número de canciones reproducidas en Springfield el lunes
monday_springfield_tracks = number_tracks('Monday', 'Springfield')
print(f'Numero de canciones reproducidas en Springfield el lunes : {monday_springfield_tracks}')

Numero de canciones reproducidas en Springfield el lunes : 15740


In [33]:
# El número de canciones reproducidas en Shelbyville el lunes
monday_shelbyville_tracks = number_tracks('Monday', 'Shelbyville')
print(f'Numero de canciones reproducidas en Shelbyville el lunes : {monday_shelbyville_tracks}')

Numero de canciones reproducidas en Shelbyville el lunes : 5614


In [34]:
# El número de canciones reproducidas en Springfield el miércoles
wednesday_springfield_tracks = number_tracks('Wednesday', 'Springfield')
print(f'Numero de canciones reproducidas en Springfield el miercoles : {wednesday_springfield_tracks}')

Numero de canciones reproducidas en Springfield el miercoles : 11056


In [35]:
# El número de canciones reproducidas en Shelbyville el miércoles
wednesday_shelbyville_tracks = number_tracks('Wednesday', 'Shelbyville')
print(f'Numero de canciones reproducidas en Shelbyville el miercoles : {wednesday_shelbyville_tracks}')

Numero de canciones reproducidas en Shelbyville el miercoles : 7003


In [36]:
# El número de canciones reproducidas en Springfield el viernes
friday_springfield_tracks = number_tracks('Friday', 'Springfield')
print(f'Numero de canciones reproducidas en Springfield el viernes : {friday_springfield_tracks}')

Numero de canciones reproducidas en Springfield el viernes : 15945


In [37]:
# El número de canciones reproducidas en Shelbyville el viernes
friday_shelbyville_tracks = number_tracks('Friday', 'Shelbyville')
print(f'Numero de canciones reproducidas en Shelbyville el viernes : {friday_shelbyville_tracks}')

Numero de canciones reproducidas en Shelbyville el viernes : 5895


**Conclusiones**

`Comenta si la hipótesis es correcta o se debe rechazar. Explica tu razonamiento. es correcta hay diferencias en cuanto a es mayor en Springfield que Shelbyville el numero de canciones reproducidas segun nos dicen los datos 

[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

`Resume aquí tus conclusiones sobre la hipótesis.` Mediante el filtrado de datos, la agrupacion de los datos nos da pie a que ambas ciudades tienen diferencias en cuantos al numero de canciones reproducidas aunque para apoyar esta hipotesis se necesita un analisis estadistico que nos complemente nuesta afirmacion

### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

[Volver a Contenidos](#back)